[libc++] Add assumption for align of begin and end pointers of vector.

fhahn opened the pull request:
Missing information about begin and end pointers of std::vector can lead to missed optimizations in LLVM.

This patch adds alignment assumptions at the point where the begin and end pointers are loaded. If the pointers would not have the same alignment, end might never get hit when incrementing begin.

See https://github.com/llvm/llvm-project/issues/101372 for a discussion of missed range check optimizations in hardened mode.

Once https://github.com/llvm/llvm-project/pull/108958 lands, the created `llvm.assume` calls for the alignment should be folded into the `load` instructions, resulting in no extra instructions after InstCombine.

fhahn committed:
[libc++] Add assumption for align of begin and end pointers of vector.

Missing information about begin and end pointers of std::vector can lead
to missed optimizations in LLVM.

See https://github.com/llvm/llvm-project/issues/101372 for a discussion
of missed range check optimizations in hardened mode.

Once https://github.com/llvm/llvm-project/pull/108958 lands, the created
`llvm.assume` calls for the alignment should be folded into the `load`
instructions, resulting in no extra instructions after InstCombine.

philnik777 commented on libcxx/include/vector:
```diff
@@ -1027,6 +1027,10 @@ private:
   }
 
   _LIBCPP_CONSTEXPR_SINCE_CXX20 _LIBCPP_HIDE_FROM_ABI void __move_assign_alloc(vector&, false_type) _NOEXCEPT {}
+
+  _LIBCPP_CONSTEXPR_SINCE_CXX20 _LIBCPP_HIDE_FROM_ABI pointer __add_alignment_assumption(pointer __p) _NOEXCEPT {
+    return static_cast<pointer>(__builtin_assume_aligned(__p, alignof(decltype(*__p))));
+  }
```
Does this work with fancy pointers? Also, shouldn't the compiler be able to infer this?

danakj commented on libcxx/include/vector:
```diff
@@ -1027,6 +1027,10 @@ private:
   _LIBCPP_CONSTEXPR_SINCE_CXX20 _LIBCPP_HIDE_FROM_ABI void __move_assign_alloc(vector&, false_type) _NOEXCEPT {}
+
+  _LIBCPP_CONSTEXPR_SINCE_CXX20 _LIBCPP_HIDE_FROM_ABI pointer __add_alignment_assumption(pointer __p) const _NOEXCEPT {
+    return static_cast<pointer>(__builtin_assume_aligned(__p, alignof(decltype(*__p))));
```
It seems that `__builtin_assume_aligned` can't be used in this way in a constexpr context since it returns `void*`:

Example:
```
../../base/containers/checked_iterators_unittest.cc:52:49: error: constexpr variable 'begin' must be initialized by a constant expression
   52 |   constexpr CheckedContiguousConstIterator<int> begin = MakeConstIter(arr, 0u);
      |                                                 ^       ~~~~~~~~~~~~~~~~~~~~~~
../../base/containers/checked_iterators.h:35:12: note: cast from 'void *' is not allowed in a constant expression in C++ standards before C++2c
   35 |     return static_cast<T*>(__builtin_assume_aligned(t, alignof(T)));
      |            ^
```

This works around the compile issue for me in Chromium with clang, though I hope it still optimizes the same.
```cc
  if (std::is_constant_evaluated()) {
    return t;
  } else {
    return static_cast<T*>(__builtin_assume_aligned(t, alignof(T)));
  }
```

I think we'd need to maintain constexpr for these methods in libc++ too, hopefully there are tests that use the iterators from a constexpr context?

ldionne replied:
We should have plenty of `constexpr` tests with iterators in `libcxx/test/std/containers/sequences/vector/iterators.pass.cpp`. Are you able to reduce the issue you were seeing a bit so we can confirm whether there is coverage for this in the test suite?

danakj replied:
https://godbolt.org/z/ndx8Pe3de shows the error, we can't static_cast from void* in constexpr

ldionne commented on libcxx/include/vector:
```diff
@@ -1027,6 +1027,10 @@ private:
 
   _LIBCPP_CONSTEXPR_SINCE_CXX20 _LIBCPP_HIDE_FROM_ABI void __move_assign_alloc(vector&, false_type) _NOEXCEPT {}
+
+  _LIBCPP_CONSTEXPR_SINCE_CXX20 _LIBCPP_HIDE_FROM_ABI pointer __add_alignment_assumption(pointer __p) const _NOEXCEPT {
```
We can make it `static`.

ldionne commented on libcxx/include/vector:
```diff
@@ -1027,6 +1027,10 @@ private:
+
+  _LIBCPP_CONSTEXPR_SINCE_CXX20 _LIBCPP_HIDE_FROM_ABI pointer __add_alignment_assumption(pointer __p) const _NOEXCEPT {
+    return static_cast<pointer>(__builtin_assume_aligned(__p, alignof(decltype(*__p))));
+  }
```
Reintroducing @philnik777 's comment:

> Does this work with fancy pointers? Also, shouldn't the compiler be able to infer this?

No, that wouldn't work AFAIU and I don't think this can work for fancy pointers (which could be implemented in very unusual ways). We may need to guard this with `if constexpr (is_pointer<pointer>::value)`.

ldionne committed:
Address review comments

ldionne committed:
Handle C++03

ldionne committed:
Add missing include

ldionne committed:
Avoid instantiating vector::begin() when declaring flat_map

Otherwise, this prevents us from being able to declare a flat_map containing
an incomplete type.

ldionne commented on libcxx/include/__flat_map/key_value_iterator.h:
```diff
@@ -15,9 +15,7 @@
 #include <__config>
```
@huixie90 Do these changes look correct to you? Is there a reason why you used `ranges::iterator_t` in the first place?

zmodem commented on libcxx/include/__vector/vector.h:
```diff
@@ -775,6 +780,17 @@ class _LIBCPP_TEMPLATE_VIS vector {
 
   _LIBCPP_CONSTEXPR_SINCE_CXX20 _LIBCPP_HIDE_FROM_ABI void __move_assign_alloc(vector&, false_type) _NOEXCEPT {}
+
+  static _LIBCPP_CONSTEXPR_SINCE_CXX20 _LIBCPP_HIDE_FROM_ABI pointer __add_alignment_assumption(pointer __p) _NOEXCEPT {
```
Similarly to https://github.com/llvm/llvm-project/pull/118837#pullrequestreview-2578321015 I think this one needs `_LIBCPP_NO_CFI` since we're potentially static_cast'ing uninitialized memory (for the end pointer).

(Apologies for the late notice, we're a bit behind on libc++.)

ldionne replied:
Ack. I'd be curious to see how this failure can be reproduced. Is it just a matter of adding a `-fsanitize=cfi` job to our pre-commit CI?

ldionne replied:
Also https://github.com/llvm/llvm-project/pull/124839 for adding NO_CFI

zmodem replied:
I haven't tried to reproduce it on the libc++ tests directly, but it should be possible. I think we're currently only using a subset of the cfi checks: `-fsanitize=cfi-vcall -fsanitize=cfi-derived-cast -fsanitize=cfi-unrelated-cast`. The annoying thing is that cfi requires (thin) lto.

winner245 commented on libcxx/include/__vector/vector.h:
```diff
@@ -775,6 +780,17 @@ class _LIBCPP_TEMPLATE_VIS vector {
   _LIBCPP_CONSTEXPR_SINCE_CXX20 _LIBCPP_HIDE_FROM_ABI void __move_assign_alloc(vector&, false_type) _NOEXCEPT {}
+
+  static _LIBCPP_CONSTEXPR_SINCE_CXX20 _LIBCPP_HIDE_FROM_ABI pointer __add_alignment_assumption(pointer __p) _NOEXCEPT {
+#ifndef _LIBCPP_CXX03_LANG
+    if constexpr (is_pointer<pointer>::value) {
```
@ldionne Are we using `constexpr if` in C++11/14 as a compiler extension here? I've also noticed several other usages of `constexpr if` in C++14 mode in libc++. 
Does this imply that we can generally use `constexpr if` within the context of C++11/14 in libc++?

ldionne replied:
Yes, it's fine to use compiler extensions in the library as long as our supported compilers implement them.
